src/t_stream.c

```diff
@@ -2259,16 +2266,28 @@ void xreadCommand(client *c) {
         /* If a group was specified, than we need to be sure that the
          * key and group actually exist. */
         if (groupname) {
-            if (o == NULL ||
-                (group = streamLookupCG(o->ptr,groupname->ptr)) == NULL)
-            {
-                addReplyErrorFormat(c, "-NOGROUP No such key '%s' or consumer "
-                                       "group '%s' in XREADGROUP with GROUP "
-                                       "option",
-                                    (char*)key->ptr,(char*)groupname->ptr);
+            if (o == NULL) {
+                addReplyErrorFormat(c, "-NOGROUP No such key '%s'"
+                                       "in XREADGROUP with GROUP "
+                                       "option",(char*)key->ptr);
                 goto cleanup;
-            }
-            groups[id_idx] = group;
+	    }
+            group = streamLookupCG(o->ptr,groupname->ptr);
+            if (group == NULL) {
+               if (!mkgroup) {
+                  addReplyErrorFormat(c, "-NOGROUP No consumer "
+                                       "group '%s' in XREADGROUP with GROUP "
+                                       "option",(char*)groupname->ptr);
+                  goto cleanup;
+               }
+               streamID id;
+               id.ms = 0;
+               id.seq = 0;
+               group = streamCreateCG(o->ptr,groupname->ptr,strlen(groupname->ptr),&id,SCG_INVALID_ENTRIES_READ);
+               server.dirty++;
+               notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",key,c->db->id);
```

Comment:
let's move the creation of the group to after argument parsing.
so that we know we're not making changes if the command is gonna fail on an error.